syntax = "proto3";

package lawnmower;

// =============================
// 基础信息类型
// =============================

// 二维坐标
message Vector2 {
    float x = 1; // 坐标X（像素单位）
    float y = 2; // 坐标Y（像素单位）
}

message Timestamp {
  uint64 server_time = 1; // 服务器时间戳（ms）
  uint32 tick = 2; // 游戏帧编号
}

// =============================
// 消息类型枚举
// =============================

enum MessageType{
  MSG_UNKNOWN = 0; // 未知信息

  // 登陆相关
  MSG_C2S_LOGIN = 1; // C2S 登陆
  MSG_S2C_LOGIN_RESULT = 2;  // S2C 登陆反馈
  MSG_C2S_HEARTBEAT = 3; // C2S 心跳信息
  MSG_S2C_HEARTBEAT = 4; // S2C 心跳信息

  // 房间相关
    // 客户端to服务器
  MSG_C2S_CREATE_ROOM = 10; // 创建房间请求
  MSG_C2S_GET_ROOM_LIST = 11; // 获取房间列表请求
  MSG_C2S_JOIN_ROOM = 12; // 加入房间请求
  MSG_C2S_LEAVE_ROOM = 13; // 离开房间请求
  MSG_C2S_SET_READY = 14; // 设置准备状态请求
  MSG_C2S_START_GAME = 15; // 开始游戏请求

    // 服务器to客户端
  MSG_S2C_CREATE_ROOM_RESULT = 20; // 创建房间反馈
  MSG_S2C_ROOM_LIST = 21; // 房间列表反馈
  MSG_S2C_JOIN_ROOM_RESULT = 22; // 加入房间反馈
  MSG_S2C_LEAVE_ROOM_RESULT = 23; // 离开房间反馈
  MSG_S2C_ROOM_UPDATE = 25; // 房间信息更新（广播)
  MSG_S2C_GAME_START = 26; // 游戏开始反馈 (广播)

  // 游戏相关
  MSG_C2S_PLAYER_INPUT = 30; // C2S 玩家信息输入
  MSG_S2C_GAME_STATE_SYNC = 40; // 广播：反馈游戏状态
  MSG_S2C_PLAYER_HURT = 41; // 广播： 玩家受伤
  MSG_S2C_ENEMY_DIED = 42; // 广播： 敌人死亡
  MSG_S2C_PLAYER_LEVEL_UP = 43; // 广播： 玩家升级
  MSG_S2C_DROPPED_ITEM = 44; // 广播： 掉落道具
  MSG_S2C_GAME_OVER = 45; // 广播： 游戏结束
}   

// =============================
// 登陆相关信息
// =============================

// 客户端 -> 服务器： 登陆请求
message C2S_Login {
    string player_name = 1; // 玩家名
}

// 服务器 -> 客户端： 登陆结果
message S2C_LoginResult {
    bool success = 1; // 登陆结果
    uint32 player_id= 2; // 玩家ID
    string message_login = 3; // 信息
}

// 客户端 -> 服务器： 心跳消息
message C2S_Heartbeat {
  uint64 timestamp = 1; // 时间戳
}

// 服务器 -> 客户端： 心跳消息
message S2C_Heartbeat {
  uint64 timestamp = 1; // 时间戳
  uint32 online_players = 2; // 在线玩家个数
}

// =============================
// 房间相关信息
// =============================

// 房间信息
message RoomInfo {
  uint32 room_id = 1; // 房间id
  string room_name = 2; // 房间名
  uint32 current_players = 3; // 当前玩家数
  uint32 max_players = 4; // 最大玩家数
  bool is_playing = 5; // 是否游戏中
  string host_name = 6; // 房主名称
}

// 房间内玩家信息
message PlayerInfo {
  uint32 player_id = 1; // 玩家id
  string player_name = 2; // 玩家姓名
  bool is_ready = 3; // 是否准备
  bool is_host = 4; // 是否房主
}

// --------------客户端->服务器--------------

// 创建房间请求
message C2S_CreateRoom {
  string room_name = 1; // 房间名
  uint32 max_players = 2; // 最大玩家数量
}

// 获取房间列表请求
message C2S_GetRoomList {
}

// 加入房间请求
message C2S_JoinRoom {
  uint32 room_id = 1; //房间id
}

// 离开房间请求
message C2S_LeaveRoom {
}

// 设置准备状态请求
message C2S_SetReady {
  bool is_ready = 1; // 是否准备
}

// 开始游戏请求（房主)
message C2S_StartGame {
}

// --------------服务器->客户端--------------

// 创建房间反馈
message S2C_CreateRoomResult {
  bool success = 1; // 是否创建成功
  uint32 room_id = 2; // 房间id
  string message_create = 3; // 信息
}

// 房间列表反馈
message S2C_RoomList {
  repeated RoomInfo rooms = 1; // 所有房间信息
}

// 加入房间反馈
message S2C_JoinRoomResult {
  bool success = 1; // 是否加入成功
  string message_join = 2; // 信息
}

// 离开房间反馈
message S2C_LeaveRoomResult {
  bool success = 1; // 是否离开成功
  string message_leave = 2; // 信息
}

// 房间信息更新(广播)
message S2C_RoomUpdate {
  uint32 room_id = 1; // 房间id
  repeated PlayerInfo players = 2; // 当前房间内所有玩家
}

// 游戏开始反馈(广播)
message S2C_GameStart {
  uint32 room_id = 1; // 房间id
  uint64 start_time = 2; // 游戏开始时间戳
}

// =============================
// 游戏相关信息
// =============================

// --------------游戏状态------------------

// 玩家状态（网络同步核心数据）
message PlayerState {
    uint32 player_id = 1;          // 玩家唯一ID
    Vector2 position = 2;          // 位置
    float rotation = 3;            // 朝向（角度，单位：度）
    int32 health = 4;              // 当前血量
    int32 max_health = 5;          // 最大血量
    uint32 level = 6;               // 等级
    uint32 exp = 7;                 // 当前经验
    uint32 exp_to_next = 8;         // 升级所需经验
    bool is_alive = 9;             // 是否存活
    uint32 attack = 10;             // 基础攻击力
    bool is_friendly = 11;         // 是否为友方（通常玩家为 true）
    uint32 role_id = 12;           // 角色ID（预留，如豌豆射手、向日葵等）
    uint32 critical_hit_rate = 13;  // 暴击率（单位：‰ 或 %，由客户端解释）
    bool has_buff = 14;            // 是否有灌注/增益效果
    uint32 buff_id = 15;           // Buff ID（0 表示无）
    uint32 attack_speed = 16;       // 攻击速度（数值越大越快，或表示攻击间隔 ms）
}

// 射弹信息（如豌豆、子弹）
message Projectile {
    uint32 speed = 1;               // 飞行速度（像素/秒）
    bool has_buff = 2;             // 是否携带 Buff 效果
    uint32 buff_id = 3;            // Buff ID
    bool is_friendly = 4;          // 是否为友方射弹
    uint32 damage = 5;              // 总伤害（服务端计算后下发）
}

// 敌人状态
message EnemyState {
    uint32 enemy_id = 1;           // 敌人实例ID（运行时生成）
    uint32 type_id = 2;            // 敌人类型ID（关联 EnemyConfig）
    Vector2 position = 3;          // 位置
    int32 health = 4;              // 当前血量
    int32 max_health = 5;          // 最大血量
    bool is_alive = 6;             // 是否存活
    uint32 wave_id = 7;            // 所属波次
    bool is_friendly = 8;          // 是否为友方（通常为 false，除非有召唤机制）
}

// 掉落物状态
message ItemState {
    uint32 item_id = 1;            // 道具实例ID
    uint32 type_id = 2;            // 道具类型ID（关联 ItemConfig）
    Vector2 position = 3;          // 位置
    bool is_picked = 4;            // 是否已被拾取
    int32 value = 5;               // 数值
}

// --------------客户端->服务器--------------

// 玩家输入
message C2S_PlayerInput {
    uint32 player_id = 1;          // 玩家唯一ID
    Vector2 move_direction = 2;    // 移动方向（归一化向量）
    bool is_attacking = 3;         // 是否正在攻击
    Timestamp input_time = 4;      // 输入时间戳（用于延迟补偿）
}

// --------------服务器->客户端--------------

// 游戏状态同步（定期发送，如每秒20次）
message S2C_GameStateSync {
  Timestamp sync_time = 1;
  repeated PlayerState players = 2;            // 变化的玩家状态
  repeated EnemyState enemies = 3;             // 变化的敌人状态
  repeated ItemState items = 4;                // 道具状态（如新掉落或被拾取）
}

// 玩家受伤广播
message S2C_PlayerHurt {
  uint32 player_id = 1;        // 受伤玩家ID
  uint32 damage = 2;           // 扣除的伤害值
  int32 remaining_health = 3;  // 受伤后的血量
  uint32 source_id = 4;        // 伤害来源ID（玩家或敌人）
}

// 敌人死亡广播
message S2C_EnemyDied {
  uint32 enemy_id = 1;          // 死亡的敌人实例ID
  uint32 killer_player_id = 2;  // 造成击杀的玩家ID
  uint32 wave_id = 3;           // 敌人所属波次
  Vector2 position = 4;         // 死亡位置，用于播放特效
}

// 玩家升级广播
message S2C_PlayerLevelUp {
  uint32 player_id = 1;      // 升级玩家ID
  uint32 new_level = 2;      // 新等级
  uint32 exp_to_next = 3;    // 升级后到下一级所需经验
}

// 道具掉落广播
message S2C_DroppedItem {
  repeated ItemState items = 1;   // 新掉落的道具列表
  uint32 source_enemy_id = 2;     // 触发掉落的敌人实例ID（无则为0）
  uint32 wave_id = 3;             // 掉落所属波次（无则为0）
}

// 玩家得分反馈
message PlayerScore {
  uint32 player_id = 1; // 玩家id
  string player_name = 2; // 玩家名称
  int32 final_level = 3; // 最终等级
  int32 kill_count = 4; // 击杀数量
  int32 damage_dealt = 5; // 总伤害
}

// 游戏结束反馈
message S2C_GameOver {
  bool victory = 1; // 是否胜利
  uint32 survive_time = 2; // 存活时间
  repeated PlayerScore scores = 3; // 所有玩家得分
}

// —————— 以下为配置类（仅客户端/服务端本地加载，不通过网络传输）——————

// 敌人配置（静态数据）
message EnemyConfig {
    uint32 type_id = 1;            // 与 EnemyState.type_id 对应
    string name = 2;               // 名称（如 "普通僵尸"）
    string texture_id = 3;         // 贴图资源ID（如 "zombie_normal"）
    string walk_anim_id = 4;       // 行走动画ID
    string die_anim_id = 5;        // 死亡动画ID
    int32 max_health = 6;          // 最大血量（用于服务端生成）
    float speed = 7;               // 移动速度（像素/秒）
    uint32 damage = 8;              // 单次伤害
    uint32 exp_reward = 9;          // 击杀后奖励经验
}

// 道具配置
message ItemConfig {
    uint32 type_id = 1;            // 与 ItemState.type_id 对应
    string name = 2;
    string texture_id = 3;         // 贴图资源ID（如 "coin"）
    string pick_effect_id = 4;     // 拾取特效ID（如 "pick_coin"）
    uint32 exp_value = 5;           // 若为经验道具，此字段有效
}

// =============================
// 网络包封装
// =============================

// 网络消息封装（通用包格式）
message Packet {
    MessageType msg_type = 1;           // 消息类型（使用 MessageType 枚举值）
    bytes payload = 2;             // 序列化后的具体消息内容
}
